airbyte-integrations/connectors/source-google-sheets/unit_tests/integration/test_source.py

```diff
@@ -0,0 +1,673 @@
+        )
+
+        http_mocker.get(
+            RequestBuilder().with_spreadsheet_id(_SPREADSHEET_ID).with_include_grid_data(False).with_alt("json").build(),
+            HttpResponse(json.dumps(find_template("invalid_permissions", __file__)), 403)
+        )
+        with pytest.raises(AirbyteTracedException) as exc_info:
+            self._source.check(Mock(), self._config)
+        assert str(exc_info.value) == (
+            "Config error: "
+        )
+
+    @HttpMocker()
+    def test_check_expected_to_read_data_from_1_sheet(self, http_mocker: HttpMocker) -> None:
+        http_mocker.post(
+            AuthBuilder.get_token_endpoint().build(),
+            HttpResponse(json.dumps(find_template("auth_response", __file__)), 200)
+        )
+
+        http_mocker.get(
+            RequestBuilder().with_spreadsheet_id(_SPREADSHEET_ID).with_include_grid_data(False).with_alt("json").build(),
+            HttpResponse(json.dumps(find_template("check_succeeded_meta", __file__)), 200)
+        )
+
+        http_mocker.get(
+            RequestBuilder().with_spreadsheet_id(_SPREADSHEET_ID).with_include_grid_data(True).with_ranges(f"{_STREAM_NAME}!1:1").with_alt("json").build(),
+            HttpResponse(json.dumps(find_template("check_wrong_range", __file__)), 200)
+        )
+
+        connection_status = self._source.check(Mock(), self._config)
+        assert connection_status.status == Status.FAILED
+        assert connection_status.message == f'Unable to read the schema of sheet a_stream_name. Error: Unexpected return result: Sheet {_STREAM_NAME} was expected to contain data on exactly 1 sheet. '
+
+    @HttpMocker()
+    def test_check_duplicated_headers(self, http_mocker: HttpMocker) -> None:
+        http_mocker.post(
+            AuthBuilder.get_token_endpoint().build(),
+            HttpResponse(json.dumps(find_template("auth_response", __file__)), 200)
+        )
+
+        http_mocker.get(
+            RequestBuilder().with_spreadsheet_id(_SPREADSHEET_ID).with_include_grid_data(False).with_alt("json").build(),
+            HttpResponse(json.dumps(find_template("check_succeeded_meta", __file__)), 200)
+        )
+
+        http_mocker.get(
+            RequestBuilder().with_spreadsheet_id(_SPREADSHEET_ID).with_include_grid_data(True).with_ranges(f"{_STREAM_NAME}!1:1").with_alt("json").build(),
+            HttpResponse(json.dumps(find_template("check_duplicate_headers", __file__)), 200)
+        )
+
+
+        connection_status = self._source.check(Mock(), self._config)
+        assert connection_status.status == Status.FAILED
+        assert connection_status.message == f"The following duplicate headers were found in the following sheets. Please fix them to continue: [sheet:{_STREAM_NAME}, headers:['header1']]"
+
+    @HttpMocker()
+    def test_given_grid_sheet_type_with_at_least_one_row_when_discover_then_return_stream(self, http_mocker: HttpMocker) -> None:
+        # spreadsheet_metadata request
```

Comment:
nit: This is done on a lot of tests. Should we extract this in a meaningful name instead of a comment? I'm also a but confused about `spreadsheet_metadata request` vs `auth_response`

Reply:
Yes, I think I left behind a few confusing comments let me fix that.

Reply:
@maxi297  I have created a few mock helpers and put some documentation in those, which I expect clarifies what they are doing.

- authorize
- get_streams
- get_schema
- get_stream_data

Also, this made the tests smaller and put the focus on the following work to move to low-code (get dynamic streams, schemas logic, etc.)